File: implement-shell-tools/cat/my_cat.js
const fs = require('fs');
const args = process.argv.slice(2);

let numberNonBlank = false;
let numberLines = false;
let files = [];

// Separate flags from filenames
for (let arg of args) {
    if (arg === '-n') {
        numberLines = true;
    } else if (arg === '-b') {
        numberNonBlank = true;
    } else {
        files.push(arg);
    }
}

let lineNumber = 1;  // start counting from 1

for (let file of files) {
    const content = fs.readFileSync(file, 'utf8');
    const lines = content.split('\n');

    for (let line of lines) {
        if (numberNonBlank) {
            if (line.trim() !== '') {
                console.log(lineNumber + '\t' + line);
                lineNumber++;
            } else {
                console.log(line);   // blank line, no number
            }
        } else if (numberLines) {
            console.log(lineNumber + '\t' + line);
            lineNumber++;
        } else {
            console.log(line);
        }
    }
}

File: implement-shell-tools/cat/my_cat.py
import sys

args = sys.argv[1:]

numberNonBlank = False
numberLines = False
files = []

for arg in args:
    if arg == '-n':
        numberLines = True
    elif arg == '-b':
        numberNonBlank = True
    else:
        files.append(arg)

lineNumber = 1

for file in files:
    f = open(file, 'r')
    content = f.read()
    f.close()
    
    lines = content.split('\n')
    
    for line in lines:
        if numberNonBlank:
            if line != '':
                print(str(lineNumber) + '\t' + line)
                lineNumber = lineNumber + 1
            else:
                print(line)
        elif numberLines:
            print(str(lineNumber) + '\t' + line)
            lineNumber = lineNumber + 1
        else:
            print(line)